### PR TITLE
🎉 Source Gitlab: Ingest All Accessible Groups

### LINKED ISSUE #11128: Improvement: Gitlab Source
## Tell us about the new connector you’d like to have
Gitlab Source

## Describe the context around this modified connector
The Gitlab source connector where you're dealing with one group with no structure and a simple list of repos. The connector support breaks down for larger teams where groups have subgroups etc.

What I would like to do:

* Use a single group-level access key that has appropriate permissions for everything within it
* Be able to leave the groups and projects field of the source config empty so that all groups and projects that are accessible to the key are ingested (i.e. not have to individually nominate each team's groups within the org group)
* Ensure that groups within groups are recursed to ensure repos that are deep in a nested structure are ingested
* Save the source configuration without receiving an error message stating: `The connection tests failed. non-json response`

## Describe the alternative you are considering or using
n/a

## Are you willing to submit a PR?
Yes

### PR DESCRIPTION
**Pull request by adamschmidt**
## What
Resolves #11128 - Allow all Gitlab groups accessible to the access token to be ingested

## How
Removes the need to specify either group ids and/or project ids in the source configuration. Where neither are specified, use the Gitlab groups API to retrieve the full list of groups that are accessible to the connector and ingest as normal.

## Recommended reading order
1. airbyte-integrations/connectors/source-gitlab/source_gitlab/source.py
2. airbyte-integrations/connectors/source-gitlab/source_gitlab/util.py

## 🚨 User Impact 🚨
User will no longer be forced to specify either a set of group ids or project ids, it should no longer fail when the user tries to save with blank in these fields. Existing configurations should not be impacted by this change.

## Pre-merge Checklist
<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [x] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret`
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [x] Documentation updated
    - [x] Connector's `README.md`
    - [x] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [x] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the new connector version is published, connector version bumped in the seed directory as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

## Tests

<details><summary><strong>Integration</strong></summary>

Simple configuration to keep the test cycle short. Configured in `airbyte-integrations/connectors/source-gitlab/integration_tests/configured_catalog.json`

```
{
  "streams": [
    {
      "stream": {
        "name": "groups",
        "json_schema": {},
        "supported_sync_modes": ["full_refresh"],
        "source_defined_primary_key": [["id"]]
      },
      "sync_mode": "full_refresh",
      "destination_sync_mode": "overwrite"
    }
  ]
}
```

Output from `python main.py read --config secrets/config.json --catalog integration_tests/configured_catalog.json`:

```
... snip ...
{"type": "LOG", "log": {"level": "INFO", "message": "Read 157 records from groups stream"}}
{"type": "LOG", "log": {"level": "INFO", "message": "Finished syncing groups"}}
{"type": "LOG", "log": {"level": "INFO", "message": "SourceGitlab runtimes:\nSyncing stream groups 0:03:11.009996"}}
{"type": "LOG", "log": {"level": "INFO", "message": "Finished syncing SourceGitlab"}}
```

</details>

### CODE REVIEW

File: airbyte-integrations/connectors/source-gitlab/source_gitlab/util.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+    while 'X-Next-Page' in r.headers and r.headers['X-Next-Page'] != '':
+        next_page = r.headers['X-Next-Page']
+        per_page = r.headers['X-Per-Page']
+        r = requests.get(f'https://{kwargs["api_url"]}/api/v4/groups?page={next_page}&per_page={per_page}', headers=headers)
+        results = r.json()
+        items = map(lambda i: i['full_path'].replace('/', '%2f'), results)
+        ids.extend(items)
+
+    return ids
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-gitlab/source_gitlab/util.py:21]:
```suggestion
    headers = kwargs["authenticator"].get_auth_header()

    ids = []
    has_next = True
    # First request params
    per_page = 50
    next_page = 1

    while has_next:
        r = requests.get(f'https://{kwargs["api_url"]}/api/v4/groups?page={next_page}&per_page={per_page}', headers=headers)
        next_page = r.headers.get('X-Next-Page')
        per_page = r.headers.get('X-Per-Page')
        results = r.json()
        items = map(lambda i: i['full_path'].replace('/', '%2f'), results)
        ids.extend(items)
        has_next = 'X-Next-Page' in r.headers and r.headers['X-Next-Page'] != ''

    return ids
```

File: airbyte-integrations/connectors/source-gitlab/source_gitlab/util.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
+import requests
+
+def get_group_list(**kwargs):
+    headers = kwargs["authenticator"].get_auth_header()
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-gitlab/source_gitlab/util.py:4]: This function is only used by the Source class, maybe transfer it to inside the class or let inside the source.py file. Don't see a separate file is helping here.

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
some changes

File: airbyte-integrations/connectors/source-gitlab/source_gitlab/util.py (hunk 2 above)
Comment (adamschmidt) [reply airbyte-integrations/connectors/source-gitlab/source_gitlab/util.py]: Moved to an internal class method on the source

File: airbyte-integrations/connectors/source-gitlab/source_gitlab/util.py (hunk 1 above)
Comment (adamschmidt) [reply airbyte-integrations/connectors/source-gitlab/source_gitlab/util.py]: Refactored. This was admittedly messy :)

**Review by marcosmarxm — APPROVED** [review]
thanks @adamschmidt I made a few changes but it's good now

**Comment by briveira** [comment]
Still seeing "non-json response" fails when adding a local GITLAB source with hundreds of projects in 0.36.7-alpha (after HOURS trying to add the new gitlab source)

**Pull request merged**